### PR TITLE
Bugfix FXIOS-5123 [v112] jump back in crash

### PR DESCRIPTION
**Pull request by OrlaM**
[FXIOS-5123](https://mozilla-hub.atlassian.net/browse/FXIOS-5123)
#12188 

Potential fix for the jump back in crash

### CODE REVIEW

File: Client/Frontend/Home/HomepageViewModel.swift [hunk 1]
```diff
@@ -175,6 +174,10 @@ class HomepageViewModel: FeatureFlaggable {
         jumpBackInViewModel.delegate = self
         messageCardViewModel.delegate = self
 
+        Task {
+            await jumpBackInAdaptor.setDelegate(delegate: jumpBackInViewModel)
```
Comment (nbhasin2) [inline Client/Frontend/Home/HomepageViewModel.swift:178]: 🔥 🤞🏼

**Review by nbhasin2 — APPROVED** [review]

File: Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift [hunk 2]
```diff
@@ -47,16 +46,22 @@ class JumpBackInDataAdaptorImplementation: JumpBackInDataAdaptor, FeatureFlaggab
         self.mainQueue = mainQueue
         self.userInitiatedQueue = userInitiatedQueue
```
Comment (lmarceau) [inline Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift:47]: I think the userInitiatedQueue can be removed

File: Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift [hunk 3]
```diff
@@ -82,85 +87,85 @@ class JumpBackInDataAdaptorImplementation: JumpBackInDataAdaptor, FeatureFlaggab
...
-        if featureFlags.isFeatureEnabled(.tabTrayGroups, checking: .buildAndUser) {
-            updateGroupsData { [weak self] in
-                self?.delegate?.didLoadNewData()
-            }
+    private func updateTabsData() async {
+        recentTabs = await updateRecentTabs()
+        if let remoteTabs = await updateRemoteTabs() {
+            createMostRecentSyncedTab(from: remoteTabs)
         }
+        recentGroups = await updateGroupsData()
```
Comment (lmarceau) [inline Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift:106]: I am writing this just to ensure we cover the case I am mentioning down here. I am unsure this still applies but better be safe than sorry. I tried getting the issue locally but so far nothing on your branch, all seems good to me  🤷‍♀️👍

--
So one of the issues I had in the past with the jbi section, is that each data type can take a different amount of time to load. For instance, we could have local tabs for jump back in, but loading the remote tabs will take a longer amount of time. The section was not updated until we had all data type (so remote tabs loaded), potentially showing outdated local tabs (some could be closed and clicking on them would crash).

To fix this problem, in the past I had `didLoadNewData` called in each data type. So any time "normal tabs" would be loaded, or "remote tabs" would be loaded, the section would update. This made sure we would show up-to-date data in that section, and avoid crashing (from PR https://github.com/mozilla-mobile/firefox-ios/pull/12290, particularly with commit https://github.com/mozilla-mobile/firefox-ios/pull/12290/commits/e1846eee491e3989e7dcd06551aa0069810ff868).

File: Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift [hunk 4]
```diff
@@ -82,85 +87,85 @@ class JumpBackInDataAdaptorImplementation: JumpBackInDataAdaptor, FeatureFlaggab
...
+    private func updateGroupsData() async -> [ASGroup<Tab>]? {
+        guard featureFlags.isFeatureEnabled(.tabTrayGroups, checking: .buildAndUser) else {
+            return nil
+        }
+        return await withCheckedContinuation { continuation in
+            SearchTermGroupsUtility.getTabGroups(
+                with: self.profile,
+                from: self.recentTabs,
+                using: .orderedDescending
+            ) { [weak self] groups, _ in
```
Comment (lmarceau) [inline Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift:128]: warning here with self
Screen Shot 2023-02-16 at 4 39 52 PM

File: Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift [hunk 5]
```diff
@@ -82,85 +87,85 @@ class JumpBackInDataAdaptorImplementation: JumpBackInDataAdaptor, FeatureFlaggab
...
+        guard hasSyncedTabFeatureEnabled() else {
+            return nil
         }
 
-        // Get cached tabs
-        profile.getCachedClientsAndTabs { [weak self] result in
-            self?.createMostRecentSyncedTab(from: result, completion: completion)
+        return await withCheckedContinuation { continuation in
+            // Get cached tabs
+            profile.getCachedClientsAndTabs { [weak self] result in
```
Comment (lmarceau) [inline Client/Frontend/Home/JumpBackIn/Data/JumpBackInDataAdaptor.swift:156]: Warning here with self too
Screen Shot 2023-02-16 at 4 39 58 PM

**Review by lmarceau — APPROVED** [review]

**Review by nbhasin2 — APPROVED** [review]

**Comment by OrlaM** [comment]
Pushed the branch to mozilla origin to force Bitrise to work and the build is green 
image

**Comment by OrlaM** [comment]
@Mergifyio backport release/v111.0

**Pull request merged**